extensions/gmsa-dc/v1/Setup-gMSA.ps1

```diff
@@ -0,0 +1,278 @@
+	
+	Install-Module powershell-yaml -Force
+
+    $CredSpecJson = Get-Item $FullPath | Select-Object @{
+        Name       = 'Name'
+        Expression = { $_.Name }
+    },
+    @{
+        Name       = 'Path'
+        Expression = { $_.FullName }
+    }
+	
+	# This section of code borrowed from https://github.com/kubernetes-sigs/windows-gmsa/blob/master/scripts/GenerateCredentialSpecResource.ps1
+    $dockerCredSpecPath = $CredSpecJson.Path
+    Sleep 2
+    $credSpecContents = Get-Content $dockerCredSpecPath | ConvertFrom-Json
+    $ManifestFile = "gmsa-cred-spec-gmsa-e2e.yml"
+	
+	# generate the k8s resource
+    $resource = [ordered]@{
+        "apiVersion" = "windows.k8s.io/v1alpha1";
+        "kind" = 'GMSACredentialSpec';
+        "metadata" = @{
+        "name" = $AccountName
+        };
+        "credspec" = $credSpecContents
+    }
+
+ConvertTo-Yaml $resource | Set-Content $ManifestFile
+
+Write-Output "K8S manifest rendered at $ManifestFile"
+}
+
+# Change to working directory
+Set-Location -Path $GMSARoot
+
+# Script Logging - Feel free to comment out if you like.
+Start-Transcript -Path "$GMSARoot\Setup-gmsa.txt" -Append
+# Making sure the AD services are up and running
+do {$ADresult = CheckADServices} while ($ADresult -eq $false)
+
+# Import Active Directory PowerShell module
+do {$ADPSresult = ImportADPS} while ($ADPSresult -eq $false)
+
+# Make sure AD server instance is listening on port
+do {$CGresult = CheckGroup} while ($CGresult -eq $false)
+
+
+if (Get-AdGroupMember -Identity "Enterprise Admins" | Select-String -Pattern "gmsa-admin" -Quiet) {
+    
+    # Add the KDS Root Key
+    $KdsRootKey = Get-KdsRootKey
+    if ($KdsRootKey -eq $null) {
+        # This command creates the KDS Root Key with a time 10 hours in the past.
+        # Active Directory will not permit use of this key until at least 10 hours has
+        # passed to allow distribution to all Domain Controllers in the domain.
+        # https://docs.microsoft.com/en-us/windows-server/security/group-managed-service-accounts/create-the-key-distribution-services-kds-root-key
+        Add-KdsRootKey –EffectiveTime ((get-date).addhours(-10))
+        # Make sure we are PAST the 10 hour mark
+        Start-Sleep -Seconds 15
+    }
+    
+    # Directory for credspecs if not already created
+    mkdir -Path C:\ProgramData\docker\credentialspecs -ErrorAction SilentlyContinue
+
+    # Import AD Module and Setup gMSA
+    New-ADServiceAccount -Name gmsa-e2e -DNSHostName gmsa-e2e.k8sgmsa.lan -PrincipalsAllowedToRetrieveManagedPassword "Domain Computers" -ServicePrincipalnames http/gmsa-e2e.k8sgmsa.lan
+
+    # Run New-CredntialSpec to provide the Yaml CredSpec
+    New-CredentialSpec -Name gmsa-e2e -AccountName gmsa-e2e
+	
+    # Copy files to make them available for download
+	Copy-Item -Path "C:\gmsa\admin.txt" -Destination "C:\inetpub\wwwroot\"
+    Copy-Item -Path "C:\gmsa\gmsa-cred-spec-gmsa-e2e.yml" -Destination "C:\inetpub\wwwroot\gmsa-cred-spec-gmsa-e2e.txt"
```

Comment:
good thing these nodes don't have an external IP :)

Reply:
Yeah.  I really couldn't figure out a way around that.  At least the password is randomized for each run!  :)